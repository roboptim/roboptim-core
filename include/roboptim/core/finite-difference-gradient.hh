// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FINITE_DIFFERENCE_GRADIENT_HH
# define ROBOPTIM_CORE_FINITE_DIFFERENCE_GRADIENT_HH
# include <stdexcept>

# include <roboptim/core/fwd.hh>
# include <roboptim/core/differentiable-function.hh>
# include <roboptim/core/portability.hh>

namespace roboptim
{
  /// \brief Default threshold for checkGradient.
  static const double finiteDifferenceThreshold = 1e-4;
  /// \brief Default epsilon for finite difference class.
  static const double finiteDifferenceEpsilon = 1e-8;

  /// \brief Exception thrown when a gradient check fails.
  template <typename T>
  class BadGradient : public std::runtime_error
  {
  public:
    ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
    (GenericDifferentiableFunction<T>);

    /// \brief Default constructor.
    BadGradient (const vector_t& x,
		 const gradient_t& analyticalGradient,
		 const gradient_t& finiteDifferenceGradient,
		 const value_type& threshold);

    virtual ~BadGradient () throw ();

    /// \brief Display the exception on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const throw ();


    /// \brief Gradient has been computed for this point.
    vector_t x_;

    /// \brief Analytical gradient.
    gradient_t analyticalGradient_;

    /// \brief Gradient computed through finite differences.
    gradient_t finiteDifferenceGradient_;

    /// \brief Maximum error.
    value_type maxDelta_;

    /// \brief Component containing the maximum error.
    size_type maxDeltaComponent_;

    /// \brief Allowed threshold.
    value_type threshold_;
  };

  /// \brief Override operator<< to handle exception display.
  ///
  /// \param o output stream used for display
  /// \param f function to be displayed
  /// \return output stream
  template <typename T>
  std::ostream& operator<< (std::ostream& o,
			    const BadGradient<T>& f);

  /// \brief Exception thrown when a Jacobian check fails.
  template <typename T>
  class BadJacobian : public std::runtime_error
  {
  public:
    ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
    (GenericDifferentiableFunction<T>);

    /// \brief Default constructor.
    BadJacobian (const vector_t& x,
		 const jacobian_t& analyticalJacobian,
		 const jacobian_t& finiteDifferenceJacobian,
		 const value_type& threshold);

    virtual ~BadJacobian () throw ();

    /// \brief Display the exception on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const throw ();


    /// \brief Jacobian has been computed for this point.
    vector_t x_;

    /// \brief Analytical Jacobian.
    gradient_t analyticalJacobian_;

    /// \brief Jacobian computed through finite differences.
    gradient_t finiteDifferenceJacobian_;

    /// \brief Maximum error.
    value_type maxDelta_;

    /// \brief Components containing the maximum error.
    size_type maxDeltaRow_;
    size_type maxDeltaCol_;

    /// \brief Allowed threshold.
    value_type threshold_;
  };

  /// \brief Override operator<< to handle exception display.
  ///
  /// \param o output stream used for display
  /// \param f function to be displayed
  /// \return output stream
  template <typename T>
  std::ostream& operator<< (std::ostream& o,
			    const BadJacobian<T>& f);

  /// \brief Contains finite difference gradients policies.
  ///
  /// Each class of this algorithm implements a finite difference
  /// gradient computation algorithm.
  namespace finiteDifferenceGradientPolicies
  {
    /// \brief Interface for the finite difference gradient policies.
    template <typename T>
    class Policy
    {
    public:
      ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
      (GenericDifferentiableFunction<T>);

      virtual void computeGradient
      (const GenericFunction<T>& adaptee,
       value_type epsilon,
       gradient_t& gradient,
       const argument_t& argument,
       size_type idFunction,
       argument_t& xEps) const throw () = 0;

      virtual void computeJacobian
      (const GenericFunction<T>& adaptee,
       value_type epsilon,
       jacobian_t& jacobian,
       const argument_t& argument,
       argument_t& xEps) const throw ();
    };

    /// \brief Fast finite difference gradient computation.
    ///
    /// Finite difference is computed using forward difference.
    template <typename T>
    class Simple : public Policy<T>
    {
    public:
      ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
      (GenericDifferentiableFunction<T>);

      void computeGradient
      (const GenericFunction<T>& adaptee,
       value_type epsilon,
       gradient_t& gradient,
       const argument_t& argument,
       size_type idFunction,
       argument_t& xEps) const throw ();
    };

    /// \brief Precise finite difference gradient computation.
    ///
    /// Finite difference is computed using five-points stencil
    /// (i.e. \f$\{x-2h, x-h, x, x+h, x+2h\}\f$).
    template <typename T>
    class FivePointsRule : public Policy<T>
    {
    public:
      ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
      (GenericDifferentiableFunction<T>);

      void computeGradient
      (const GenericFunction<T>& adaptee,
       value_type epsilon,
       gradient_t& gradient,
       const argument_t& argument,
       size_type idFunction,
       argument_t& xEps) const throw ();
    };
  } // end of namespace policy.


  /// \addtogroup roboptim_function
  /// @{

  /// \brief Compute automatically a gradient with finite differences.
  ///
  /// Finite difference gradient is a method to approximate a function's
  /// gradient. It is particularly useful in RobOptim to avoid the need
  /// to compute the analytical gradient manually.
  ///
  /// This class takes a Function as its input and wraps it into a derivable
  /// function.
  ///
  /// The one dimensional formula is:
  /// \f[f'(x)\approx {f(x+\epsilon)-f(x)\over \epsilon}\f]
  /// where \f$\epsilon\f$ is a constant given when calling the class
  /// constructor.
  template <typename T, typename FdgPolicy>
  class GenericFiniteDifferenceGradient
    : public GenericDifferentiableFunction<T>,
      protected FdgPolicy
  {
  public:
    ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
    (GenericDifferentiableFunction<T>);

    /// \brief Instantiate a finite differences gradient.
    ///
    /// Instantiate a derivable function that will wraps a non
    /// derivable function and compute automatically its gradient
    /// using finite differences.
    /// \param f function that will e wrapped
    /// \param e epsilon used in finite difference computation
    GenericFiniteDifferenceGradient
    (const GenericFunction<T>& f,
     value_type e = finiteDifferenceEpsilon) throw ();
    ~GenericFiniteDifferenceGradient () throw ();

  protected:
    virtual void impl_compute (result_t&, const argument_t&) const throw ();
    virtual void impl_gradient (gradient_t&,
                                const argument_t& argument,
                                size_type = 0) const throw ();
    virtual void impl_jacobian (jacobian_t& jacobian,
                                const argument_t& argument) const throw ();

    /// \brief Reference to the wrapped function.
    const GenericFunction<T>& adaptee_;

    //// \brief Epsilon used in finite differences computation.
    const value_type epsilon_;

    mutable argument_t xEps_;
  };

  /// \brief Check if a gradient is valid.
  ///
  /// Check if a gradient is valid by comparing the distance between its
  /// gradient and an automatically computed finite differences gradient.
  /// \param function function that will be checked
  /// \param functionId function id in split representation
  /// \param x point where the gradient will be evaluated
  /// \param threshold maximum tolerated error
  /// \return true if valid, false if not
  template <typename T>
  bool
  checkGradient
  (const GenericDifferentiableFunction<T>& function,
   typename GenericDifferentiableFunction<T>::size_type functionId,
   const typename GenericDifferentiableFunction<T>::vector_t& x,
   typename GenericDifferentiableFunction<T>::value_type threshold =
   finiteDifferenceThreshold)
    throw ();

  template <typename T>
  void
  checkGradientAndThrow
  (const GenericDifferentiableFunction<T>& function,
   typename GenericDifferentiableFunction<T>::size_type functionId,
   const typename GenericDifferentiableFunction<T>::vector_t& x,
   typename GenericDifferentiableFunction<T>::value_type threshold =
   finiteDifferenceThreshold)
    throw (BadGradient<T>);

  /// \brief Check if a Jacobian is valid.
  ///
  /// Check if a Jacobian is valid by comparing the distance between the
  /// matrix and an automatically computed finite differences Jacobian.
  /// \param function function that will be checked
  /// \param x point where the Jacobian will be evaluated
  /// \param threshold maximum tolerated error
  /// \return true if valid, false if not
  template <typename T>
  bool
  checkJacobian
  (const GenericDifferentiableFunction<T>& function,
   const typename GenericDifferentiableFunction<T>::vector_t& x,
   typename GenericDifferentiableFunction<T>::value_type threshold =
   finiteDifferenceThreshold)
    throw ();

  template <typename T>
  void
  checkJacobianAndThrow
  (const GenericDifferentiableFunction<T>& function,
   const typename GenericDifferentiableFunction<T>::vector_t& x,
   typename GenericDifferentiableFunction<T>::value_type threshold =
   finiteDifferenceThreshold)
    throw (BadJacobian<T>);

  /// Example shows finite differences gradient use.
  /// \example finite-difference-gradient.cc

  /// @}

} // end of namespace roboptim

# include <roboptim/core/finite-difference-gradient.hxx>
#endif //! ROBOPTIM_CORE_FINITE_DIFFERENCE_GRADIENT_HH
