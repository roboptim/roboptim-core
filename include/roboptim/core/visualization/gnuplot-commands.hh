// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <string>
# include <iostream>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      /// \addtogroup roboptim_visualization
      /// @{

      /// \brief Gnuplot command.
      ///
      /// A command represents a Gnuplot command.
      /// The representation of the command as a string
      /// can be accessed through the #command() method.
      class ROBOPTIM_DLLAPI Command
      {
      public:
	/// \brief Make a command from a string.
	explicit Command (std::string cmd) throw ();

	~Command () throw ();

	/// \brief Retrieve the command as a string.
	const std::string& command () const throw ();
      protected:
	/// \brief Store Gnuplot command.
	std::string command_;
      };

      /// \brief Make a Gnuplot comment.
      ROBOPTIM_DLLAPI Command comment (const char*) throw ();

      /// \brief Make a Gnuplot comment.
      template <typename T>
      ROBOPTIM_DLLAPI Command comment (const T& content) throw ()
      {
          // Note: we do not use boost::lexical_cast because the << operators
          // need to be in the std:: or boost:: namespaces. As a result, if we
          // try to add a comment with an Eigen matrix, it will not be printed
          // in the RobOptim way. Thus, we stick to stringstream (for now).
          std::stringstream ss;
          ss << "# " << content;
          return Command (ss.str ());
      }

      /// \brief Make a Gnuplot set command.
      ///
      /// Make a Gnuplot set command (assign a new
      /// value to a variable in Gnuplot).
      /// \param var variable that will be set
      /// \param val new variable value
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command set (const char* var, const char* val = "") throw ();

      /// \brief Make a Gnuplot unset command.
      ///
      /// Make a Gnuplot unset command (destroy a variable).
      /// \param var variable that will be unset
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command unset (const char* var) throw ();

      /// \brief Make a Gnuplot show command.
      ///
      /// Make a Gnuplot show command (show a variable value).
      /// \param var variable that will be shown
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command show (const char* var) throw ();

      /// \brief Make a Gnuplot clear command.
      ///
      /// Make a Gnuplot clear command (clear the screen).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command clear () throw ();

      /// \brief Make a Gnuplot pwdcommand.
      ///
      /// Make a Gnuplot pwd command (show the current directory).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command pwd () throw ();

      /// \brief Make a Gnuplot quit command.
      ///
      /// Make a Gnuplot quit command (quit Gnuplot).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command quit () throw ();

      /// \brief Make a Gnuplot replot command.
      ///
      /// Make a Gnuplot replot command (redraw previously defined curves).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command replot () throw ();

      /// \brief Make a Gnuplot reread command.
      ///
      /// Make a Gnuplot reread command (reread data files).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command reread () throw ();

      /// \brief Make a Gnuplot reset command.
      ///
      /// Make a Gnuplot replot command (reset Gnuplot state).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command reset () throw ();

      /// \brief Make a Gnuplot cd command.
      ///
      /// Make a Gnuplot cd command (change current directory).
      /// \param dir new directory
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command cd (const char* dir) throw ();

      /// \brief Make a Gnuplot help command.
      ///
      /// Make a Gnuplot help command (display Gnuplot help).
      /// \param topic help topic
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command help (const char* topic = "") throw ();

      /// @}

    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
