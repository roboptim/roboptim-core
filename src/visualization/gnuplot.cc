// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "roboptim/core/config.hh"
#include "config.h"

#include <iostream>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/indent.hh>

namespace roboptim
{
  namespace visualization
  {
    Gnuplot::Gnuplot (bool with_header) throw ()
    {
      using namespace gnuplot;
      push_command (comment ("!/usr/bin/gnuplot"));
      if(with_header) push_command (comment (" Generated by " PACKAGE_NAME));
    }

    Gnuplot::~Gnuplot () throw ()
    {
    }

    void
    Gnuplot::push_command (gnuplot::Command cmd) throw ()
    {
      commands_.push_back (cmd);
    }

    std::ostream&
    Gnuplot::print (std::ostream& o) const throw ()
    {
      typedef std::vector<gnuplot::Command>::const_iterator citer_t;

      for (citer_t it = commands_.begin (); it != commands_.end (); ++it)
	o << it->command () << std::endl;
      return o;
    }

    Gnuplot&
    Gnuplot::operator << (gnuplot::Command cmd) throw ()
    {
      push_command (cmd);
      return *this;
    }

    std::ostream&
    operator<< (std::ostream& o, const Gnuplot& gp)
    {
      return gp.print (o);
    }
  } // end of namespace visualization.
} // end of namespace roboptim.
