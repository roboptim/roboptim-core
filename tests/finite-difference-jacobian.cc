// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/finite-difference-gradient.hh>
#include <roboptim/core/indent.hh>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/visualization/gnuplot-function.hh>

using namespace roboptim;
using namespace roboptim::visualization;
using namespace roboptim::visualization::gnuplot;

// Define a function with a correct Jacobian.
template <typename T>
struct FGood : public GenericDifferentiableFunction<T>
{
  ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
  (GenericDifferentiableFunction<T>);

  FGood () : GenericDifferentiableFunction<T>
             (2, 2, "x * x + x * y + 2 * y, 3 * x * x * y")
  {}

  void impl_compute (result_t& result,
		     const argument_t& argument) const throw ()
  {
    result (0) =   argument[0] * argument[0]
      + argument[0] * argument[1]
      +          2. * argument[1];
    result (1) = 3. * argument[0] * argument[0] * argument[1];
  }

  void impl_gradient (gradient_t& grad, const argument_t& argument,
		      size_type) const throw ();

  void impl_jacobian (jacobian_t& jacobian,
                      const argument_t& argument) const throw ();

};


template <typename T>
void
FGood<T>::impl_gradient
(gradient_t&, const argument_t&, size_type) const throw ()
{
  // This method should not be called since we provide impl_jacobian
  assert(0);
}


template <>
void
FGood<EigenMatrixSparse>::impl_jacobian
(jacobian_t& jacobian, const argument_t& argument) const throw ()
{
  jacobian.setZero ();
  jacobian.insert(0,0) = 2. * argument[0] + argument[1];
  jacobian.insert(0,1) =      argument[0] + 2.;
  jacobian.insert(1,0) = 6. * argument[0] * argument[1];
  jacobian.insert(1,1) = 3. * argument[0] * argument[0];
}

template <typename T>
void
FGood<T>::impl_jacobian
(jacobian_t& jacobian, const argument_t& argument) const throw ()
{
  jacobian.setZero ();
  jacobian(0,0) = 2. * argument[0] + argument[1];
  jacobian(0,1) =      argument[0] + 2.;
  jacobian(1,0) = 6. * argument[0] * argument[1];
  jacobian(1,1) = 3. * argument[0] * argument[0];
}


// Define a function with a bad Jacobian.
template <typename T>
struct FBad : public GenericDifferentiableFunction<T>
{
  ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
  (GenericDifferentiableFunction<T>);

  FBad () : GenericDifferentiableFunction<T>
            (2, 2, "x * x + x * y + 2 * y, 3 * x * x * y")
  {}

  void impl_compute (result_t& result,
		     const argument_t& argument) const throw ()
  {
    result (0) =   argument[0] * argument[0]
      + argument[0] * argument[1]
      +          2. * argument[1];
    result (1) = 3. * argument[0] * argument[0] * argument[1];
  }

  void impl_gradient (gradient_t& grad, const argument_t& argument,
		      size_type) const throw ();

  void impl_jacobian (jacobian_t& jacobian,
                      const argument_t& argument) const throw ();
};

template <typename T>
void
FBad<T>::impl_gradient
(gradient_t&, const argument_t&, size_type) const throw ()
{
  // This method should not be called since we provide impl_jacobian
  assert(0);
}


template <>
void
FBad<EigenMatrixSparse>::impl_jacobian
(jacobian_t& jacobian, const argument_t& argument) const throw ()
{
  jacobian.setZero ();
  jacobian.insert(0,0) = 2. * argument[0] + argument[1];
  jacobian.insert(0,1) =      argument[0] + 42.;
  jacobian.insert(1,0) = 6. * argument[0] * argument[1];
  jacobian.insert(1,1) = 3. * argument[0] * argument[0];
}

template <typename T>
void
FBad<T>::impl_jacobian
(jacobian_t& jacobian, const argument_t& argument) const throw ()
{
  jacobian.setZero ();
  jacobian(0,0) = 2. * argument[0] + argument[1];
  jacobian(0,1) =      argument[0] + 42.;
  jacobian(1,0) = 6. * argument[0] * argument[1];
  jacobian(1,1) = 3. * argument[0] * argument[0];
}


template <typename T>
void displayJacobian
(boost::shared_ptr<boost::test_tools::output_test_stream> output,
 const GenericDifferentiableFunction<T>&,
 const typename GenericDifferentiableFunction<T>::vector_t&);

template <typename T>
void displayJacobian
(boost::shared_ptr<boost::test_tools::output_test_stream> output,
 const GenericDifferentiableFunction<T>& function,
 const typename GenericDifferentiableFunction<T>::vector_t& x)
{
  GenericFiniteDifferenceGradient<T> fdfunction (function);
  typename GenericFiniteDifferenceGradient<T>::jacobian_t jac =
    function.jacobian (x);
  typename GenericFiniteDifferenceGradient<T>::jacobian_t fdjac =
    fdfunction.jacobian (x);

  (*output) << "#" << jac << std::endl
            << "#" << fdjac << std::endl;
}


template <>
void displayJacobian<roboptim::EigenMatrixSparse>
(boost::shared_ptr<boost::test_tools::output_test_stream> output,
 const GenericDifferentiableFunction<roboptim::EigenMatrixSparse>& function,
 const GenericDifferentiableFunction<roboptim::EigenMatrixSparse>::
 vector_t& x)
{
  GenericFiniteDifferenceGradient<roboptim::EigenMatrixSparse>
    fdfunction (function);
  GenericFiniteDifferenceGradient<roboptim::EigenMatrixSparse>::
    jacobian_t jac = function.jacobian (x);
  GenericFiniteDifferenceGradient<roboptim::EigenMatrixSparse>::
    jacobian_t fdjac = fdfunction.jacobian (x);

  (*output) << "#" << sparse_to_dense(jac) << std::endl
	    << "#" << sparse_to_dense(fdjac) << std::endl;
}

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

typedef boost::mpl::list< ::roboptim::EigenMatrixDense,
			  ::roboptim::EigenMatrixSparse> functionTypes_t;

BOOST_AUTO_TEST_CASE_TEMPLATE (finite_difference_jacobian, T, functionTypes_t)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("finite-difference-jacobian");

  FGood<T> fg;
  FBad<T> fb;

  typename FGood<T>::vector_t x (2);

  for (double i = -10.; i < 10.; i += 1.)
    {
      x[0] = i;
      x[1] = 2. * i + 0.5;

      (*output) << "#Check Jacobian at x = " << x << std::endl;

      (*output) << "# Good" << std::endl;
      displayJacobian (output, fg, x);

      (*output) << "# Bad" << std::endl;
      displayJacobian (output, fb, x);

      BOOST_CHECK (checkJacobian (fg, x));
      BOOST_CHECK (! checkJacobian (fb, x));
    }

  std::cout << output->str () << std::endl;

  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
