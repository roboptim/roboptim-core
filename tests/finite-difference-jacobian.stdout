#Check Jacobian at x = [2](-10,-19.5)
# Good
#[1,2]((-39.5,-8))
#[1,2]((   -39.5,-7.99998))
# Bad
#[1,2](( 22,-12))
#[1,2]((   -39.5,-7.99998))
#Check Jacobian at x = [2](-9,-17.5)
# Good
#[1,2]((-35.5,-7))
#[1,2]((   -35.5,-6.99999))
# Bad
#[1,2](( 24,-11))
#[1,2]((   -35.5,-6.99999))
#Check Jacobian at x = [2](-8,-15.5)
# Good
#[1,2]((-31.5,-6))
#[1,2]((   -31.5,-6.00001))
# Bad
#[1,2](( 26,-10))
#[1,2]((   -31.5,-6.00001))
#Check Jacobian at x = [2](-7,-13.5)
# Good
#[1,2]((-27.5,-5))
#[1,2]((   -27.5,-4.99999))
# Bad
#[1,2]((28,-9))
#[1,2]((   -27.5,-4.99999))
#Check Jacobian at x = [2](-6,-11.5)
# Good
#[1,2]((-23.5,-4))
#[1,2]((-23.5,-4))
# Bad
#[1,2]((30,-8))
#[1,2]((-23.5,-4))
#Check Jacobian at x = [2](-5,-9.5)
# Good
#[1,2]((-19.5,-3))
#[1,2]((-19.5,-3))
# Bad
#[1,2]((32,-7))
#[1,2]((-19.5,-3))
#Check Jacobian at x = [2](-4,-7.5)
# Good
#[1,2]((-15.5,-2))
#[1,2]((-15.5,-2))
# Bad
#[1,2]((34,-6))
#[1,2]((-15.5,-2))
#Check Jacobian at x = [2](-3,-5.5)
# Good
#[1,2]((-11.5,-1))
#[1,2]((    -11.5,-0.999999))
# Bad
#[1,2]((36,-5))
#[1,2]((    -11.5,-0.999999))
#Check Jacobian at x = [2](-2,-3.5)
# Good
#[1,2]((-7.5,0))
#[1,2]((       -7.5,5.92119e-08))
# Bad
#[1,2]((38,-4))
#[1,2]((       -7.5,5.92119e-08))
#Check Jacobian at x = [2](-1,-1.5)
# Good
#[1,2]((-3.5,1))
#[1,2]((-3.5,1))
# Bad
#[1,2]((40,-3))
#[1,2]((-3.5,1))
#Check Jacobian at x = [2](0,0.5)
# Good
#[1,2]((0.5,2))
#[1,2]((0.5,2))
# Bad
#[1,2]((42,-2))
#[1,2]((0.5,2))
#Check Jacobian at x = [2](1,2.5)
# Good
#[1,2]((4.5,3))
#[1,2]((4.5,3))
# Bad
#[1,2]((44,-1))
#[1,2]((4.5,3))
#Check Jacobian at x = [2](2,4.5)
# Good
#[1,2]((8.5,4))
#[1,2]((8.5,4))
# Bad
#[1,2]((46,0))
#[1,2]((8.5,4))
#Check Jacobian at x = [2](3,6.5)
# Good
#[1,2]((12.5,5))
#[1,2]((12.5,5))
# Bad
#[1,2]((48,1))
#[1,2]((12.5,5))
#Check Jacobian at x = [2](4,8.5)
# Good
#[1,2]((16.5,6))
#[1,2]((16.5,6))
# Bad
#[1,2]((50,2))
#[1,2]((16.5,6))
#Check Jacobian at x = [2](5,10.5)
# Good
#[1,2]((20.5,7))
#[1,2]((   20.5,7.00001))
# Bad
#[1,2]((52,3))
#[1,2]((   20.5,7.00001))
#Check Jacobian at x = [2](6,12.5)
# Good
#[1,2]((24.5,8))
#[1,2]((   24.5,7.99999))
# Bad
#[1,2]((54,4))
#[1,2]((   24.5,7.99999))
#Check Jacobian at x = [2](7,14.5)
# Good
#[1,2]((28.5,9))
#[1,2]((   28.5,8.99999))
# Bad
#[1,2]((56,5))
#[1,2]((   28.5,8.99999))
#Check Jacobian at x = [2](8,16.5)
# Good
#[1,2]((32.5,10))
#[1,2]((   32.5,9.99999))
# Bad
#[1,2]((58,6))
#[1,2]((   32.5,9.99999))
#Check Jacobian at x = [2](9,18.5)
# Good
#[1,2]((36.5,11))
#[1,2]((36.5,11))
# Bad
#[1,2]((60,7))
#[1,2]((36.5,11))

#Check Jacobian at x = [2](-10,-19.5)
# Good
#[1,2]((-39.5,-8))
#[1,2]((   -39.5,-7.99998))
# Bad
#[1,2](( 22,-12))
#[1,2]((   -39.5,-7.99998))
#Check Jacobian at x = [2](-9,-17.5)
# Good
#[1,2]((-35.5,-7))
#[1,2]((   -35.5,-6.99999))
# Bad
#[1,2](( 24,-11))
#[1,2]((   -35.5,-6.99999))
#Check Jacobian at x = [2](-8,-15.5)
# Good
#[1,2]((-31.5,-6))
#[1,2]((   -31.5,-6.00001))
# Bad
#[1,2](( 26,-10))
#[1,2]((   -31.5,-6.00001))
#Check Jacobian at x = [2](-7,-13.5)
# Good
#[1,2]((-27.5,-5))
#[1,2]((   -27.5,-4.99999))
# Bad
#[1,2]((28,-9))
#[1,2]((   -27.5,-4.99999))
#Check Jacobian at x = [2](-6,-11.5)
# Good
#[1,2]((-23.5,-4))
#[1,2]((-23.5,-4))
# Bad
#[1,2]((30,-8))
#[1,2]((-23.5,-4))
#Check Jacobian at x = [2](-5,-9.5)
# Good
#[1,2]((-19.5,-3))
#[1,2]((-19.5,-3))
# Bad
#[1,2]((32,-7))
#[1,2]((-19.5,-3))
#Check Jacobian at x = [2](-4,-7.5)
# Good
#[1,2]((-15.5,-2))
#[1,2]((-15.5,-2))
# Bad
#[1,2]((34,-6))
#[1,2]((-15.5,-2))
#Check Jacobian at x = [2](-3,-5.5)
# Good
#[1,2]((-11.5,-1))
#[1,2]((    -11.5,-0.999999))
# Bad
#[1,2]((36,-5))
#[1,2]((    -11.5,-0.999999))
#Check Jacobian at x = [2](-2,-3.5)
# Good
#[1,2]((-7.5,0))
#[1,2]((       -7.5,5.92119e-08))
# Bad
#[1,2]((38,-4))
#[1,2]((       -7.5,5.92119e-08))
#Check Jacobian at x = [2](-1,-1.5)
# Good
#[1,2]((-3.5,1))
#[1,2]((-3.5,1))
# Bad
#[1,2]((40,-3))
#[1,2]((-3.5,1))
#Check Jacobian at x = [2](0,0.5)
# Good
#[1,2]((0.5,2))
#[1,2]((0.5,2))
# Bad
#[1,2]((42,-2))
#[1,2]((0.5,2))
#Check Jacobian at x = [2](1,2.5)
# Good
#[1,2]((4.5,3))
#[1,2]((4.5,3))
# Bad
#[1,2]((44,-1))
#[1,2]((4.5,3))
#Check Jacobian at x = [2](2,4.5)
# Good
#[1,2]((8.5,4))
#[1,2]((8.5,4))
# Bad
#[1,2]((46,0))
#[1,2]((8.5,4))
#Check Jacobian at x = [2](3,6.5)
# Good
#[1,2]((12.5,5))
#[1,2]((12.5,5))
# Bad
#[1,2]((48,1))
#[1,2]((12.5,5))
#Check Jacobian at x = [2](4,8.5)
# Good
#[1,2]((16.5,6))
#[1,2]((16.5,6))
# Bad
#[1,2]((50,2))
#[1,2]((16.5,6))
#Check Jacobian at x = [2](5,10.5)
# Good
#[1,2]((20.5,7))
#[1,2]((   20.5,7.00001))
# Bad
#[1,2]((52,3))
#[1,2]((   20.5,7.00001))
#Check Jacobian at x = [2](6,12.5)
# Good
#[1,2]((24.5,8))
#[1,2]((   24.5,7.99999))
# Bad
#[1,2]((54,4))
#[1,2]((   24.5,7.99999))
#Check Jacobian at x = [2](7,14.5)
# Good
#[1,2]((28.5,9))
#[1,2]((   28.5,8.99999))
# Bad
#[1,2]((56,5))
#[1,2]((   28.5,8.99999))
#Check Jacobian at x = [2](8,16.5)
# Good
#[1,2]((32.5,10))
#[1,2]((   32.5,9.99999))
# Bad
#[1,2]((58,6))
#[1,2]((   32.5,9.99999))
#Check Jacobian at x = [2](9,18.5)
# Good
#[1,2]((36.5,11))
#[1,2]((36.5,11))
# Bad
#[1,2]((60,7))
#[1,2]((36.5,11))

